### PR TITLE
[tests][python][scikit-learn] New unit tests and maintenance

### PR DESCRIPTION
**Pull request by a-wozniakowski**
This pull request:
* Includes new [multioutput](https://github.com/scikit-learn/scikit-learn/blob/fd237278e/sklearn/multioutput.py) tests for independent subtasks and chaining.
* Includes a [RandomizedSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html) test.
* Updates dataset parameters to eliminate the [FutureWarning](https://github.com/scikit-learn/scikit-learn/blob/fd237278e895b42abe8d8d09105cbb82dc2cbba7/sklearn/utils/validation.py#L39) in accordance with [PEP 3102](https://www.python.org/dev/peps/pep-3102/).
* Updates the [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html) test such that the refit best classifier is scored on the test data-- this is also done in the RandomizedSearchCV test.

### CODE REVIEW

File: tests/python_package_test/test_sklearn.py [hunk 1]
```diff
@@ -227,25 +231,107 @@ def test_grid_search(self):
...
+        X, y = make_multilabel_classification(n_classes=3, random_state=0)
+        y = y.astype(str)
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        clf = MultiOutputClassifier(estimator=lgb.LGBMClassifier(n_estimators=10))
+        clf.fit(X_train, y_train)
+        score = clf.score(X_test, y_test)
+        self.assertGreaterEqual(score, 0.2)
+        self.assertLessEqual(score, 1.)
+        for classifier in clf.estimators_:
+            self.assertIsInstance(classifier.booster_, lgb.Booster)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:282]:
```suggestion
        for classifier in clf.estimators_:
            self.assertIsInstance(classifier, lgb.LGBMClassifier)
            self.assertIsInstance(classifier.booster_, lgb.Booster)
```

File: tests/python_package_test/test_sklearn.py [hunk 2]
```diff
@@ -227,25 +231,107 @@ def test_grid_search(self):
...
+        X, y = bunch['data'], bunch['target']
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        reg = MultiOutputRegressor(estimator=lgb.LGBMRegressor(n_estimators=10))
+        reg.fit(X_train, y_train)
+        y_pred = reg.predict(X_test)
+        _, score, _ = mse(y_test, y_pred)
+        self.assertGreaterEqual(score, 0.2)
+        self.assertLessEqual(score, 120.0)
+        for regressor in reg.estimators_:
+            self.assertIsInstance(regressor.booster_, lgb.Booster)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:297]:
```suggestion
        for regressor in reg.estimators_:
            self.assertIsInstance(regressor, lgb.LGBMRegressor)
            self.assertIsInstance(regressor.booster_, lgb.Booster)
```

File: tests/python_package_test/test_sklearn.py [hunk 3]
```diff
@@ -227,25 +231,107 @@ def test_grid_search(self):
...
+
+    def test_classifier_chain(self):
+        X, y = make_multilabel_classification(n_classes=3, random_state=0)
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        clf = ClassifierChain(base_estimator=lgb.LGBMClassifier(n_estimators=10), order=[2, 0, 1])
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:302]: Fix `random_state` parameter.

File: tests/python_package_test/test_sklearn.py [hunk 4]
```diff
@@ -227,25 +231,107 @@ def test_grid_search(self):
...
+        clf.fit(X_train, y_train)
+        score = clf.score(X_test, y_test)
+        self.assertGreaterEqual(score, 0.2)
+        self.assertLessEqual(score, 1.)
+        for classifier in clf.estimators_:
+            self.assertIsInstance(classifier.booster_, lgb.Booster)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:308]:
```suggestion
        for classifier in clf.estimators_:
            self.assertIsInstance(classifier, lgb.LGBMClassifier)
            self.assertIsInstance(classifier.booster_, lgb.Booster)
```

File: tests/python_package_test/test_sklearn.py [hunk 5]
```diff
@@ -227,25 +231,107 @@ def test_grid_search(self):
...
+
+    def test_classifier_chain(self):
+        X, y = make_multilabel_classification(n_classes=3, random_state=0)
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        clf = ClassifierChain(base_estimator=lgb.LGBMClassifier(n_estimators=10), order=[2, 0, 1])
+        clf.fit(X_train, y_train)
+        score = clf.score(X_test, y_test)
+        self.assertGreaterEqual(score, 0.2)
+        self.assertLessEqual(score, 1.)
+        for classifier in clf.estimators_:
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:307]: Maybe we can assert length of `classes_` and `order_` attributes as well after fitting?

File: tests/python_package_test/test_sklearn.py [hunk 6]
```diff
@@ -227,25 +231,107 @@ def test_grid_search(self):
...
+            self.assertIsInstance(classifier.booster_, lgb.Booster)
+
+    # sklearn < 0.23 does not have as_frame parameter
+    @unittest.skipIf(sk_version < '0.23.0', 'scikit-learn version is less than 0.23')
+    def test_regressor_chain(self):
+        bunch = load_linnerud(as_frame=True)  # returns a Bunch instance
+        X, y = bunch['data'], bunch['target']
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        reg = RegressorChain(base_estimator=lgb.LGBMRegressor(n_estimators=10), order=[2, 0, 1])
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:316]: Fix `random_state` parameter.

File: tests/python_package_test/test_sklearn.py [hunk 7]
```diff
@@ -227,25 +231,107 @@ def test_grid_search(self):
...
+        reg.fit(X_train, y_train)
+        y_pred = reg.predict(X_test)
+        _, score, _ = mse(y_test, y_pred)
+        self.assertGreaterEqual(score, 0.2)
+        self.assertLessEqual(score, 120.0)
+        for regressor in reg.estimators_:
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:322]: Maybe we can assert `order_` attribute as well after fitting?

File: tests/python_package_test/test_sklearn.py [hunk 8]
```diff
@@ -227,25 +231,107 @@ def test_grid_search(self):
...
+            self.assertIsInstance(regressor.booster_, lgb.Booster)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:323]:
```suggestion
        for regressor in reg.estimators_:
            self.assertIsInstance(regressor, lgb.LGBMRegressor)
            self.assertIsInstance(regressor.booster_, lgb.Booster)
```

**Comment by a-wozniakowski** [comment]
@StrikerRUS thanks for your feedback! I implemented the requested changes, and I wrote the grid/random searches such that there is an explicit train/validation/test split.

Also, scikit-learn supports ```**fit_params``` in ```MultiOutputClassifier``` and ```MultiOutputRegressor```; see issue [#15953](https://github.com/scikit-learn/scikit-learn/issues/15953). This seems partial progress towards issue [#524](https://github.com/microsoft/LightGBM/issues/524), which is contained in the feature request hub [#2302](https://github.com/microsoft/LightGBM/issues/2302).

File: tests/python_package_test/test_sklearn.py [hunk 9]
```diff
@@ -214,38 +217,157 @@ def test_stacking_regressor(self):
...
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1,
+                                                            random_state=42)
+        clf = MultiOutputClassifier(estimator=lgb.LGBMClassifier(n_estimators=10))
+        clf.fit(X_train, y_train)
+        score = clf.score(X_test, y_test)
+        self.assertGreaterEqual(score, 0.2)
+        self.assertLessEqual(score, 1.)
+        _classes = np.ones(shape=2 * n_outputs, dtype=str)
+        _classes[0::2] = str(0)  # make binary classification str pattern
+        self.assertTrue(all(_classes == np.concatenate(clf.classes_)))
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:297]:
```suggestion
        np.testing.assert_array_equal(np.tile(np.unique(y_train), n_outputs),
                                      np.concatenate(clf.classes_))
```

File: tests/python_package_test/test_sklearn.py [hunk 10]
```diff
@@ -214,38 +217,157 @@ def test_stacking_regressor(self):
...
+        order = [2, 0, 1]
+        clf = ClassifierChain(base_estimator=lgb.LGBMClassifier(n_estimators=10),
+                              order=order, random_state=42)
+        clf.fit(X_train, y_train)
+        score = clf.score(X_test, y_test)
+        self.assertGreaterEqual(score, 0.2)
+        self.assertLessEqual(score, 1.)
+        _classes = np.ones(shape=2 * n_outputs, dtype=int)
+        _classes[0::2] = 0  # make binary classification int pattern
+        self.assertTrue(all(_classes == np.concatenate(clf.classes_)))
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:336]:
```suggestion
        np.testing.assert_array_equal(np.tile(np.unique(y_train), n_outputs),
                                      np.concatenate(clf.classes_))
```

File: tests/python_package_test/test_sklearn.py [hunk 11]
```diff
@@ -214,38 +217,157 @@ def test_stacking_regressor(self):
...
+        self.assertTrue(order == clf.order_)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:337]:
```suggestion
        self.assertListEqual(order, clf.order_)
```

File: tests/python_package_test/test_sklearn.py [hunk 12]
```diff
@@ -214,38 +217,157 @@ def test_stacking_regressor(self):
...
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+        order = [2, 0, 1]
+        reg = RegressorChain(base_estimator=lgb.LGBMRegressor(n_estimators=10), order=order,
+                             random_state=42)
+        reg.fit(X_train, y_train)
+        y_pred = reg.predict(X_test)
+        _, score, _ = mse(y_test, y_pred)
+        self.assertGreaterEqual(score, 0.2)
+        self.assertLessEqual(score, 120.)
+        self.assertTrue(order == reg.order_)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:356]:
```suggestion
        self.assertListEqual(order, reg.order_)
```

**Pull request merged**